### ISSUE TITLE
setState in componentWillMount causes TypeError when rendering with renderToString and preceded by a component that renders with renderToString

### ISSUE THREAD

**Issue by PKJedi** [facebook/react#6423]
Calling `setState` in `componentWillMount` of a component causes the error

```
TypeError: Cannot read property '_currentElement' of null
    at ReactCompositeComponentMixin._updateRenderedComponent
```

when rendering server-side with `ReactDOMServer.renderToString` and preceded by a component that renders content using `ReactDOMServer.renderToString` or `ReactDOMServer.renderToStaticMarkup`.

Causes TypeError with 0.14.8 and 15.0.0-rc.2:
https://github.com/PKJedi/react-dom-server-test/blob/master/test.js

Works with 0.13:
https://github.com/PKJedi/react-dom-server-test/blob/master/0.13/test.js

**Comment by gaearon** [comment]
Do we even support this? If we don't we should have a good early error message.

**Comment by Merri** [comment]
Yes, it is a supported feature [according to the docs](https://facebook.github.io/react/docs/component-specs.html):

> componentWillMount
> Invoked once, both on the client and server, immediately before the initial rendering occurs. If you call setState within this method, render() will see the updated state and will be executed only once despite the state change.

**Comment by Aweary** [comment]
I've been looking at this issue this morning and it seems that it stems, at least partially, from the fact that `ReactDefaultBatchingStrategy` is being when the first child component rendered has a call to `renderToStaticMarkup` or `renderToString`. 

If you switch the order and render it second, then `ReactServerBatchingStrategy` (which is a noop) is used and the error isn't thrown.

**Comment by Aweary** [comment]
So it looks like by the time `enqueueUpdate` is called, the `batchingStrategy` is already reset to `ReactDefaultBatchingStrategy`, even though the transaction that wrapped the `ReactReconciler.mountComponent` call was pulled from the `ReactServerRenderingTransaction` object pool.

Since `ReactServerRenderingTransaction.batchedUpdates` is a noop, and `ReactDefaultBatchingStrategy` is not, it's incorrectly attempting to apply the batched updates when it shouldn't.

**Comment by jimfb** [comment]
cc @spicyj

**Comment by Aweary** [comment]
I still have a rather naive understanding of how `ReactUpdates` works, but would it be reasonable to just enqueue any `injectBatchingStrategy` calls when there is still a transaction that hasn't been released, instead of applying it immediately? That way any actions invoked in `transaction.perform` would finish will the batching strategy they expected?

**Comment by spicyj** [comment]
IIRC we don't support nested server rendering at the moment and have another issue about it somewhere.

**Comment by Aweary** [comment]
@spicyj I opened a PR https://github.com/facebook/react/pull/7033 with a fix that resolves this issue, let me know what you think.
